# Vehicles_Messy

In [ ]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows',None)

import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('../data/vehicles_messy.csv')
df = data.copy()
#df.head()

In [3]:
df.shape

(37843, 83)

# Valores nulos

### 1. Nulos en cylinders

In [4]:
# Columns con al menos un valor NaN
df.columns[df.isna().any()].tolist()

['cylinders',
 'displ',
 'drive',
 'eng_dscr',
 'trany',
 'guzzler',
 'trans_dscr',
 'tCharger',
 'sCharger',
 'atvType',
 'fuelType2',
 'rangeA',
 'evMotor',
 'mfrCode',
 'c240Dscr',
 'c240bDscr',
 'startStop']

In [5]:
#df[df['cylinders'].isna()][['cylinders', 'fuelType', 'fuelType1']]

In [6]:
df.loc[[21413,21414,21506]]

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
21413  13.733750         0.0        0.0        0.0      22      0.0        0   
21414  14.330870         0.0        0.0        0.0      21      0.0        0   
21506  18.311667         0.0        0.0        0.0      15      0.0        0   

       cityA08U  cityCD  cityE  cityUF  co2  co2A  co2TailpipeAGpm  \
21413       0.0     0.0    0.0     0.0   -1    -1              0.0   
21414       0.0     0.0    0.0     0.0   -1    -1              0.0   
21506       0.0     0.0    0.0     0.0   -1    -1              0.0   

       co2TailpipeGpm  comb08  comb08U  combA08  combA08U  combE  combinedCD  \
21413      370.291667      24      0.0        0       0.0    0.0         0.0   
21414      386.391304      23      0.0        0       0.0    0.0         0.0   
21506      493.722222      18      0.0        0       0.0    0.0         0.0   

       combinedUF  cylinders  displ             drive  engId  eng_dscr  \
21413         0.0        NaN    NaN     4-Wheel Drive      0       NaN   
21414         0.0        NaN    NaN     4-Wheel Drive      0  CA model   
21506         0.0        NaN    1.3  Rear-Wheel Drive      0       NaN   

       feScore  fuelCost08  fuelCostA08 fuelType         fuelType1  ghgScore  \
21413       -1        1400            0  Regular  Regular Gasoline        -1   
21414       -1        1450            0  Regular  Regular Gasoline        -1   
21506       -1        1850            0  Regular  Regular Gasoline        -1   

       ghgScoreA  highway08  highway08U  highwayA08  highwayA08U  highwayCD  \
21413         -1         28         0.0           0          0.0        0.0   
21414         -1         27         0.0           0          0.0        0.0   
21506         -1         22         0.0           0          0.0        0.0   

       highwayE  highwayUF  hlv  hpv     id  lv2  lv4    make     model  \
21413       0.0        0.0    0    0  29472    0    0  Subaru  RX Turbo   
21414       0.0        0.0    0    0  29473    0    0  Subaru  RX Turbo   
21506       0.0        0.0    0    0  29557    0    0   Mazda      RX-7   

      mpgData  phevBlended  pv2  pv4  range  rangeCity  rangeCityA  rangeHwy  \
21413       N        False    0    0      0        0.0         0.0       0.0   
21414       N        False    0    0      0        0.0         0.0       0.0   
21506       N        False    0    0      0        0.0         0.0       0.0   

       rangeHwyA         trany    UCity  UCityA  UHighway  UHighwayA  \
21413        0.0  Manual 5-spd  27.4000     0.0   38.8000        0.0   
21414        0.0  Manual 5-spd  26.9000     0.0   37.1000        0.0   
21506        0.0  Manual 5-spd  19.0473     0.0   31.1094        0.0   

                VClass  year  youSaveSpend guzzler trans_dscr tCharger  \
21413  Subcompact Cars  1985          -250     NaN        NaN        T   
21414  Subcompact Cars  1985          -500     NaN        NaN        T   
21506      Two Seaters  1986         -2500     NaN        NaN      NaN   

      sCharger atvType fuelType2 rangeA evMotor mfrCode c240Dscr  charge240b  \
21413      NaN     NaN       NaN    NaN     NaN     NaN      NaN         0.0   
21414      NaN     NaN       NaN    NaN     NaN     NaN      NaN         0.0   
21506      NaN     NaN       NaN    NaN     NaN     NaN      NaN         0.0   

      c240bDscr                     createdOn                    modifiedOn  \
21413       NaN  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013   
21414       NaN  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013   
21506       NaN  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013   

      startStop  phevCity  phevHwy  phevComb  
21413       NaN         0        0         0  
21414       NaN         0        0         0  
21506       NaN         0        0         0

In [7]:
# Siendo solo tres se añaden a mano.
df.loc[21413, 'cylinders'] = 4
df.loc[21414, 'cylinders'] = 4
df.loc[21506, 'cylinders'] = 2

In [8]:
# Solo quedan elect. no tienen cilindros se ponen 0
df.cylinders.replace({np.NaN : 0},inplace=True)

In [9]:
#Comprobación de que todos los electricos tienen 0 cilindros
len(df[df['cylinders'] == 0]) == len(df[df["fuelType"] == "Electricity"])

True

### 2. Columnas con más de 10.000 nulos

In [10]:
df.isna().sum()[df.isna().sum() > 16_000]

guzzler       35562
trans_dscr    22796
tCharger      32657
sCharger      37177
atvType       34771
fuelType2     36435
rangeA        36440
evMotor       37281
mfrCode       30818
c240Dscr      37806
c240bDscr     37807
startStop     31705
dtype: int64

In [11]:
df = df.drop(columns=df.columns[df.isna().sum() > 16_000])

In [12]:
df.shape

(37843, 71)

Solo había una con 15k. el resto son a partir de 30k.
Se eliminan a partir de 16k.

### 3.seguimos bajando los nulos

In [13]:
# Columns con al menos un valor NaN
df.columns[df.isna().any()].tolist()

['displ', 'drive', 'eng_dscr', 'trany']

In [14]:
# Ya que la mayoría de electricos son Automatic (A1). Lo aplicamos para todos
# df[df["fuelType"] == "Electricity"]["trany"] # Presenta los datos de electricity y trany

In [15]:
df.trany = np.where(df.fuelType == "Electricity", "Automatic (A1)", df.trany)

In [16]:
df[df['trany'].isna()][["model","trany", 'cylinders', 'fuelType', 'fuelType1']]

model trany  cylinders fuelType         fuelType1
19097  F150 Pickup 2WD   NaN        8.0  Regular  Regular Gasoline
19128   C25 Pickup 2WD   NaN        6.0  Regular  Regular Gasoline

In [17]:
# ambos son auto pero con different gear ratio. Se sobreescribem
df.loc[19097, 'trany'] = "Automatic (variable gear ratios)"
df.loc[19128, 'trany'] = "Automatic (variable gear ratios)"

### 4. Nos cargamos las eng description substituyendo por Unknown

In [18]:
df.isna().sum()[df.isna().sum() > 0]

displ         120
drive        1189
eng_dscr    15403
dtype: int64

In [19]:
len(df.eng_dscr.unique())

542

In [20]:
df.eng_dscr.replace({np.NaN : "Unknown"},inplace=True)

### 5. Atacamos al drive

In [21]:
df.isna().sum()[df.isna().sum() > 0]

displ     120
drive    1189
dtype: int64

In [22]:
list(df.drive.unique())

['Rear-Wheel Drive',
 'Front-Wheel Drive',
 '4-Wheel or All-Wheel Drive',
 nan,
 '2-Wheel Drive',
 'All-Wheel Drive',
 '4-Wheel Drive',
 'Part-time 4-Wheel Drive',
 '2-Wheel Drive, Front']

In [23]:
def unit_drive (string):
    if pd.notna(string):
        string = string.lower()
        if "4" in string:
            return "4WD"
        elif "all" in string:
            return "4WD"
        else:
            return "2WD"
    else:
        return "Unknown"
                

In [24]:
df.drive = df.drive.apply(unit_drive)

Nos cargamos los NaN del drive simplificando el contenido en 2WD 4WD y Unknown con la función planteada

### 6. ultimos nulos en displ

In [25]:
df.isna().sum()[df.isna().sum() > 0]

displ    120
dtype: int64

dos opciones:
- rellenar con media, mediana, moda...
- cargarnoslo

relleno con la media para probar el método

In [26]:
mediana = df.displ.median()

In [27]:
df.displ.replace({np.NaN : mediana}, inplace = True)

In [28]:
df.isna().sum()[df.isna().sum() > 0]

Series([], dtype: int64)

### No más valores nulos¡¡¡

### Limpiamos aun más la columna trany

In [29]:
def transmission (string):
    string = string.lower()
    if "man" in string:
        return "Manual"
    else: 
        return "Automatic"

In [30]:
df.trany = df.trany.apply(transmission)

In [31]:
df.trany.unique()

array(['Manual', 'Automatic'], dtype=object)

### Nos vamos a deshacer de las columnas con baja varianza

In [32]:
# Try
per90 = df.barrels08.quantile(0.9)
per90

23.543571428571425

In [33]:
# Try
mina = df.barrelsA08.min()
mina

0.0

In [34]:
df.info(verbose = False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37843 entries, 0 to 37842
Columns: 71 entries, barrels08 to phevComb
dtypes: bool(1), float64(32), int64(27), object(11)
memory usage: 20.2+ MB


In [35]:
low_variance = []

for col in df.columns:
    if df[col].dtype != "object":
        if df[col].dtype != "bool":
            per90 = df[col].quantile(0.9)
            minimum = df[col].min()
            if per90 == minimum:
                low_variance.append(col)

print(low_variance)

['barrelsA08', 'charge120', 'charge240', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF', 'co2A', 'co2TailpipeAGpm', 'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'fuelCostA08', 'ghgScoreA', 'highwayA08', 'highwayA08U', 'highwayCD', 'highwayE', 'highwayUF', 'range', 'rangeCity', 'rangeCityA', 'rangeHwy', 'rangeHwyA', 'UCityA', 'UHighwayA', 'charge240b', 'phevCity', 'phevHwy', 'phevComb']


In [36]:
df.drop(low_variance , axis = 1 ,inplace=True)

In [37]:
df.shape

(37843, 38)

## Análisis

Ahora tenemos un df legible sin nulos y con datos limpios.
Podemos inicar el análisis y las correlaciones

Lo primero vamos a revisar el consumo ciudad vs autopista

In [38]:
df.head()

barrels08  city08  city08U  co2  co2TailpipeGpm  comb08  comb08U  \
0  15.695714      19      0.0   -1      423.190476      21      0.0   
1  29.964545       9      0.0   -1      807.909091      11      0.0   
2  12.207778      23      0.0   -1      329.148148      27      0.0   
3  29.964545      10      0.0   -1      807.909091      11      0.0   
4  17.347895      17      0.0   -1      467.736842      19      0.0   

   cylinders  displ drive  engId    eng_dscr  feScore  fuelCost08 fuelType  \
0        4.0    2.0   2WD   9011       (FFS)       -1        1600  Regular   
1       12.0    4.9   2WD  22020   (GUZZLER)       -1        3050  Regular   
2        4.0    2.2   2WD   2100       (FFS)       -1        1250  Regular   
3        8.0    5.2   2WD   2850     Unknown       -1        3050  Regular   
4        4.0    2.2   4WD  66031  (FFS,TRBO)       -1        2150  Premium   

          fuelType1  ghgScore  highway08  highway08U  hlv  hpv     id  lv2  \
0  Regular Gasoline        -1         25         0.0    0    0      1    0   
1  Regular Gasoline        -1         14         0.0    0    0     10    0   
2  Regular Gasoline        -1         33         0.0   19   77    100    0   
3  Regular Gasoline        -1         12         0.0    0    0   1000    0   
4  Premium Gasoline        -1         23         0.0    0    0  10000    0   

   lv4        make                model mpgData  phevBlended  pv2  pv4  \
0    0  Alfa Romeo   Spider Veloce 2000       Y        False    0    0   
1    0     Ferrari           Testarossa       N        False    0    0   
2    0       Dodge              Charger       Y        False    0    0   
3    0       Dodge  B150/B250 Wagon 2WD       N        False    0    0   
4   14      Subaru     Legacy AWD Turbo       N        False    0   90   

       trany    UCity  UHighway           VClass  year  youSaveSpend  \
0     Manual  23.3333   35.0000      Two Seaters  1985         -1250   
1     Manual  11.0000   19.0000      Two Seaters  1985         -8500   
2     Manual  29.0000   47.0000  Subcompact Cars  1985           500   
3  Automatic  12.2222   16.6667             Vans  1985         -8500   
4     Manual  21.0000   32.0000     Compact Cars  1993         -4000   

                      createdOn                    modifiedOn  
0  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013  
1  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013  
2  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013  
3  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013  
4  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013

In [39]:
manual_city = (df[df.trany == "Manual"].UCity.mean())
auto_city = (df[df.trany == "Automatic"].UCity.mean())

man_hig = (df[df.trany == "Manual"].UHighway.mean())
auto_hig = (df[df.trany == "Automatic"].UHighway.mean())

In [40]:
print("Manual consume city: ",manual_city)
print("Auto consume city: ",auto_city)
print("\n")
print("Manual consume highway: ",man_hig)
print("Auto consume highway: ",auto_hig)

Manual consume city:  23.803615061748328
Auto consume city:  21.995187990099787


Manual consume highway:  35.39287307288724
Auto consume highway:  32.75594702993636


Vemos que los coches automáticos tanto en ciudad como en autopista tienen un menor consumo. Si bien esto puede deberse a la presencia de los coches eléctricos entre los automáticos.
Veamos que pasaría al quitar los eléctricos

In [41]:
len(df[df.fuelType == "Electricity"])

120

In [42]:
df_el = df[df['fuelType'] != 'Electricity']

Creamos un nuevo data frame sin coches eléctricos y volvemos a analizar

In [43]:
manual_city_el = (df_el[df_el.trany == "Manual"].UCity.mean())
auto_city_el = (df_el[df_el.trany == "Automatic"].UCity.mean())
man_hig_el = (df_el[df_el.trany == "Manual"].UHighway.mean())
auto_hig_el = (df_el[df_el.trany == "Automatic"].UHighway.mean())

In [44]:
print("Consume without electric cars:")
print("\n")
print("Manual consume city: ",manual_city_el)
print("Auto consume city: ",auto_city_el)
print("\n")
print("Manual consume highway: ",man_hig_el)
print("Auto consume highway: ",auto_hig_el)

Consume without electric cars:


Manual consume city:  23.803615061748328
Auto consume city:  21.462607780058416


Manual consume highway:  35.39287307288724
Auto consume highway:  32.33731868635036


Vemos que las diferencias entre los resultados son mínimas. Es posible debido a que tan solo había 120 coches eléctricos en el data frame. En cualquier caso sigue estando claro que los coches automáticos tienen un mejor consumo que los manuales.

### Cleaning years

In [45]:
df.year.info()

<class 'pandas.core.series.Series'>
RangeIndex: 37843 entries, 0 to 37842
Series name: year
Non-Null Count  Dtype
--------------  -----
37843 non-null  int64
dtypes: int64(1)
memory usage: 295.8 KB


In [46]:
#df.year = pd.to_datetime(df.year)

In [47]:
df.year.info()

<class 'pandas.core.series.Series'>
RangeIndex: 37843 entries, 0 to 37842
Series name: year
Non-Null Count  Dtype
--------------  -----
37843 non-null  int64
dtypes: int64(1)
memory usage: 295.8 KB


In [48]:
len(df.year.unique())

34

# Outliers

In [49]:
#Run this code
stats = data.describe().transpose()
stats['IQR'] = stats['75%'] - stats['25%']
#stats.head()

In [50]:
outliers = pd.DataFrame(columns=data.columns)

for col in stats.index:
    iqr = stats.at[col,'IQR']
    cutoff = iqr * 1.5
    lower = stats.at[col,'25%'] - cutoff
    upper = stats.at[col,'75%'] + cutoff
    results = data[(data[col] < lower) | 
                   (data[col] > upper)].copy()
    results['Outlier'] = col
    outliers = outliers.append(results)
    

# outliers.head()

In [51]:
# Columnas en comón
len(set(df.columns) & set(outliers.Outlier.unique()))

23

In [52]:
# Columnas que no coinciden
#set(df.columns) | set(outliers.Outlier.unique())

# Createdon to simplify

In [53]:
#df.createdOn.unique()

In [54]:
def conversion(texto):
    month = ['jan','feb','mar','apr','may','jun','jul','aug','sep','oct','nov','dec']
    year = []
    days = []

    for i in range(1,32):
        #i = int(i)
        i = str(i)
        days.append(i)

    #print(days)


    for i in range(2000,2020):
        i = str(i)
        year.append(i)

    texto = texto.lower()
    texto = texto.split()
    #print(texto)

    fecha = []

    for i in texto:
        if i in year:
            fecha.append(i)

    for i in texto:
        if i in month:
            fecha.append(i)

    for i in texto:
        try:
            i = int(i)
            i = str(i)
        except:
            pass
        if i in days:
            fecha.append(i)

    return "-".join(fecha)

In [55]:
df["Created"] = df.createdOn.apply(conversion)
df["Modified"] = df.modifiedOn.apply(conversion)
df = df.drop(columns=["createdOn", "modifiedOn"])

In [56]:
df.head(1)

barrels08  city08  city08U  co2  co2TailpipeGpm  comb08  comb08U  \
0  15.695714      19      0.0   -1      423.190476      21      0.0   

   cylinders  displ drive  engId eng_dscr  feScore  fuelCost08 fuelType  \
0        4.0    2.0   2WD   9011    (FFS)       -1        1600  Regular   

          fuelType1  ghgScore  highway08  highway08U  hlv  hpv  id  lv2  lv4  \
0  Regular Gasoline        -1         25         0.0    0    0   1    0    0   

         make               model mpgData  phevBlended  pv2  pv4   trany  \
0  Alfa Romeo  Spider Veloce 2000       Y        False    0    0  Manual   

     UCity  UHighway       VClass  year  youSaveSpend     Created    Modified  
0  23.3333      35.0  Two Seaters  1985         -1250  2013-jan-1  2013-jan-1